File: 0078-subsets/0078-subsets.c
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
 void helper(int *nums,int numsSize,int index,int *tempsubset,int tempsize,int **result,int* returnSize,int **returnColumnSizes){
    result[*returnSize]=malloc(tempsize*sizeof(int));

    memcpy(result[*returnSize], tempsubset, tempsize * sizeof(int));
    (*returnColumnSizes)[*returnSize]=tempsize;
    (*returnSize)++;

    for(int i=index;i<numsSize;i++){
        tempsubset[tempsize]=nums[i];
        helper(nums,numsSize,i+1,tempsubset,tempsize+1,result,returnSize,returnColumnSizes);
    }
    
 }
int** subsets(int* nums, int numsSize, int* returnSize, int** returnColumnSizes) {
    *returnSize=0;
    int totalsubsets=1<<numsSize;
    int** result=(int**)malloc(totalsubsets * sizeof(int*));

    *returnColumnSizes = (int*)malloc(totalsubsets * sizeof(int));
    int* tempsubset=(int*)malloc(numsSize*sizeof(int)); 

    helper(nums,numsSize,0,tempsubset,0,result,returnSize,returnColumnSizes);

    
    return result; 
}

File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> sub = new ArrayList<>();
        int n = nums.length;
        helper(0,n,nums,ans,sub);
        return ans;
    }

    private void helper(int i,int n,int[] nums,List<List<Integer>> ans,List<Integer> sub)
    {
        if(i == n){
            if(!ans.contains(sub)){
                ans.add(new ArrayList<>(sub));
            }
            return;
        }
        sub.add(nums[i]);
        helper(i+1,n,nums,ans,sub);
        sub.remove(sub.size()-1);
        helper(i+1,n,nums,ans,sub);
    }
}